pkg/pfcpsim/session/pdr_builder.go

```diff
@@ -0,0 +1,142 @@
+package session
+
+import (
+	"net"
+
+	"github.com/wmnsk/go-pfcp/ie"
+)
+
+type pdrBuilder struct {
+	precedence uint32
+	method     IEMethod
+	sdfFilter  string
+	id         uint16
+	teid       uint32
+	farID      uint32
+	sessQerID  uint32
+	appQerID   uint32
+
+	ueAddress string
+	n3Address string
+
+	direction direction
+}
+
+func NewPDRBuilder() *pdrBuilder {
+	return &pdrBuilder{}
+}
+
+func (b *pdrBuilder) WithPrecedence(precedence uint32) *pdrBuilder {
+	b.precedence = precedence
+	return b
+}
+
+func (b *pdrBuilder) WithSDFFilter(filter string) *pdrBuilder {
+	b.sdfFilter = filter
+	return b
+}
+
+func (b *pdrBuilder) WithID(id uint16) *pdrBuilder {
+	b.id = id
+	return b
+}
+
+func (b *pdrBuilder) WithTEID(teid uint32) *pdrBuilder {
+	b.teid = teid
+	return b
+}
+
+func (b *pdrBuilder) WithMEthod(method IEMethod) *pdrBuilder {
+	b.method = method
+	return b
+}
+
+func (b *pdrBuilder) WithN3Address(n3Address string) *pdrBuilder {
+	b.n3Address = n3Address
+	return b
+}
+
+func (b *pdrBuilder) WithUEAddress(ueAddress string) *pdrBuilder {
+	b.ueAddress = ueAddress
+	return b
+}
+
+func (b *pdrBuilder) WithRulesIDs(farID uint32, sessionQERID uint32, appQERID uint32) *pdrBuilder {
```

Comment:
The PFCP spec does not differentiate between sessQERID and appQERID. I believe it's something Aether-specific.

I propose to split this method into: `WithFARID()`, `WithQERID`.  The `pdrBuilder` should store a list of QERs.

Reply:
Done. pdrBuilder now stores QERIDs and FARIDs as slices. I used the signatures `AddQERID(qerID uint32)` and `AddFARID(farID uint32)` to store new `ie.QERID`s `ie.FARID`s, respectively

Reply:
Thanks, but we should not allow to specify more than one FAR ID for a PDR, see Table 7.5.2.2-1 in the PFCP spec. It is allowed to pass several QER IDs though.